File: Data Structure & Algorithm/Double Point/en/Left & Right/167. Two Sum II - Input Array Is Sorted/167. Two Sum II - Input Array Is Sorted.ts
// 167. Two Sum II - Input Array Is Sorted

// Given a 1-indexed array of integers numbers that is already sorted in non-decreasing order, find two numbers such that they add up to a specific target number. Let these two numbers be numbers[index1] and numbers[index2] where 1 <= index1 < index2 <= numbers.length.
// Return the indices of the two numbers, index1 and index2, added by one as an integer array [index1, index2] of length 2.
// The tests are generated such that there is exactly one solution. You may not use the same element twice.
// Your solution must use only constant extra space.

// Example 1:
// Input: numbers = [2,7,11,15], target = 9
// Output: [1,2]
// Explanation: The sum of 2 and 7 is 9. Therefore, index1 = 1, index2 = 2. We return [1, 2].

// Example 2:
// Input: numbers = [2,3,4], target = 6
// Output: [1,3]
// Explanation: The sum of 2 and 4 is 6. Therefore index1 = 1, index2 = 3. We return [1, 3].

// Example 3:
// Input: numbers = [-1,0], target = -1
// Output: [1,2]
// Explanation: The sum of -1 and 0 is -1. Therefore index1 = 1, index2 = 2. We return [1, 2].

// fast & slow double point
export function twoSum(numbers: number[], target: number): number[] {
  let slow = 0;
  let fast = slow + 1;
  while (slow < numbers.length) {
    while (fast < numbers.length) {
      if (numbers[slow] + numbers[fast] === target) {
        return [slow + 1, fast + 1];
      }
      fast++;
    }
    slow++;
    fast = slow + 1;
  }

  return [-1, -1];
}

// left and right double point
export function twoSum1(numbers: number[], target: number): number[] {
  let left = 0;
  let right = numbers.length - 1;
  while (left < right) {
    if (numbers[left] + numbers[right] === target) {
      return [left + 1, right + 1];
    } else if (numbers[left] + numbers[right] < target) {
      left++;
    } else {
      right--;
    }
  }

  return [-1, -1];
}

// test
const res = twoSum([-1, 1, 0, 15], -1);
const res1 = twoSum1([-1, 1, 0, 15], -1);
console.log(res);
console.log(res1);

File: Data Structure & Algorithm/Data Structure & Algorithm [Old]/Algorithm/Two Points/en/Fast & Slow/167. Two Sum II - Input Array Is Sorted/167. Two Sum II - Input Array Is Sorted.py
# 167. Two Sum II - Input Array Is Sorted

# Given a 1-indexed array of integers numbers that is already sorted in non-decreasing order, find two numbers such that they add up to a specific target number. Let these two numbers be numbers[index1] and numbers[index2] where 1 <= index1 < index2 <= numbers.length.
# Return the indices of the two numbers, index1 and index2, added by one as an integer array [index1, index2] of length 2.
# The tests are generated such that there is exactly one solution. You may not use the same element twice.
# Your solution must use only constant extra space.

# Example 1:
# Input: numbers = [2,7,11,15], target = 9
# Output: [1,2]
# Explanation: The sum of 2 and 7 is 9. Therefore, index1 = 1, index2 = 2. We return [1, 2].

# Example 2:
# Input: numbers = [2,3,4], target = 6
# Output: [1,3]
# Explanation: The sum of 2 and 4 is 6. Therefore index1 = 1, index2 = 3. We return [1, 3].

# Example 3:
# Input: numbers = [-1,0], target = -1
# Output: [1,2]
# Explanation: The sum of -1 and 0 is -1. Therefore index1 = 1, index2 = 2. We return [1, 2].

from typing import List


class Solution:

    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        slow = 0
        fast = slow + 1
        while slow < len(numbers):
            while fast < len(numbers):
                if numbers[slow] + numbers[fast] == target:
                    return [slow + 1, fast + 1]
                fast = fast + 1

            slow = slow + 1
            fast = slow + 1
        return [-1, -1]

    def twoSum1(self, numbers: List[int], target: int) -> List[int]:
        low = 0
        high = len(numbers) - 1
        while low < high:
            total = numbers[low] + numbers[high]
            if total == target:
                return [low + 1, high + 1]
            elif total < target:
                low += 1
            else:
                high -= 1

        return [-1, -1]


# test
solution = Solution()
res = solution.twoSum([2, 7, 11, 15], 9)
res1 = solution.twoSum1([2, 7, 11, 15], 9)
print(res)
print(res1)
